registry/storage/driver/swift/swift.go

```diff
@@ -0,0 +1,533 @@
+	segmentsContainer := d.getSegmentsContainer()
+
+	getSegment := func() string {
+		return d.swiftPath(path) + "/" + fmt.Sprintf("%016d", partNumber)
+	}
+
+	max := func(a int64, b int64) int64 {
+		if a > b {
+			return a
+		}
+		return b
+	}
+
+	info, _, err := d.Conn.Object(d.Container, d.swiftPath(path))
+	if err != nil {
+		if swiftErr, ok := err.(*swift.Error); ok && swiftErr.StatusCode == 404 {
+			// Create a object manifest
+			if err := d.createParentFolders(path); err != nil {
+				return bytesRead, err
+			}
+			manifest, err := d.createManifest(path)
+			if err != nil {
+				return bytesRead, parseError(path, err)
+			}
+			manifest.Close()
+		} else {
+			return bytesRead, parseError(path, err)
+		}
+	} else {
+		// The manifest already exists. Get all the segments
+		currentLength = info.Bytes
+		segments, err = d.getAllSegments(segmentsContainer, path)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+	}
+
+	// First, we skip the existing segments that are not modified by this call
+	for i := range segments {
+		if offset < cursor+segments[i].Bytes {
+			break
+		}
+		cursor += segments[i].Bytes
+		partNumber++
+	}
+
+	// We reached the end of the file but we haven't reached 'offset' yet
+	// Therefore we add blocks of zeros
+	if offset >= currentLength {
+		for offset-currentLength >= chunkSize {
+			// Insert a block a zero
+			d.Conn.ObjectPut(segmentsContainer, getSegment(),
+				bytes.NewReader(zeroBuf), false, "",
+				d.getContentType(), nil)
+			currentLength += chunkSize
+			partNumber++
+		}
+
+		cursor = currentLength
+		paddingReader = bytes.NewReader(zeroBuf)
+	} else {
+		// Offset is inside the current segment : we need to read the
+		// data from the beginning of the segment to offset
+		paddingReader, _, err = d.Conn.ObjectOpen(segmentsContainer, getSegment(), false, nil)
+		if err != nil {
+			return bytesRead, parseError(getSegment(), err)
+		}
+	}
+
+	multi := io.MultiReader(
+		io.LimitReader(paddingReader, offset-cursor),
+		io.LimitReader(reader, chunkSize-(offset-cursor)),
+	)
+
+	for {
+		currentSegment, err := d.Conn.ObjectCreate(segmentsContainer, getSegment(), false, "", d.getContentType(), nil)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+
+		n, err := io.Copy(currentSegment, multi)
+		if err != nil {
+			return bytesRead, parseError(path, err)
+		}
+
+		if n < chunkSize {
+			// We wrote all the data
+			if cursor+n < currentLength {
+				// Copy the end of the chunk
+				headers := make(swift.Headers)
+				headers["Range"] = "bytes=" + strconv.FormatInt(cursor+n, 10) + "-" + strconv.FormatInt(cursor+chunkSize, 10)
+				file, _, err := d.Conn.ObjectOpen(d.Container, d.swiftPath(path), false, headers)
+				if err != nil {
+					return bytesRead, parseError(path, err)
+				}
+				if _, err := io.Copy(currentSegment, file); err != nil {
+					return bytesRead, parseError(path, err)
+				}
+				file.Close()
+			}
+			if n > 0 {
+				currentSegment.Close()
```

Comment:
From the godocs ["You MUST call Close() on it and you MUST check the error return from Close()"](http://godoc.org/github.com/ncw/swift#Connection.ObjectCreate)